xmake/rules/c++/modules/modules_support/builder.lua

```diff
@@ -0,0 +1,377 @@
+--!A cross-platform build utility based on Lua
+--
+-- Licensed under the Apache License, Version 2.0 (the "License");
+-- you may not use this file except in compliance with the License.
+-- You may obtain a copy of the License at
+--
+--     http://www.apache.org/licenses/LICENSE-2.0
+--
+-- Unless required by applicable law or agreed to in writing, software
+-- distributed under the License is distributed on an "AS IS" BASIS,
+-- WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+-- See the License for the specific language governing permissions and
+-- limitations under the License.
+--
+-- Copyright (C) 2015-present, TBOOX Open Source Group.
+--
+-- @author      ruki, Arthapz
+-- @file        common.lua
+--
+
+-- imports
+import("core.base.json")
+import("core.base.option")
+import("private.async.buildjobs")
+import("core.tool.compiler")
+import("core.project.config")
+import("core.project.depend")
+import("utils.progress")
+import("compiler_support")
+import("dependency_scanner")
+
+-- build target modules
+function _build_modules(target, sourcebatch, modules, opt)
+    local objectfiles = dependency_scanner.sort_modules_by_dependencies(sourcebatch.objectfiles, modules)
+
+    _builder(target).populate_module_map(target, modules)
+
+    -- build modules
+    for _, objectfile in ipairs(objectfiles) do
+        local module = modules[objectfile]
+        if not module then
+            goto CONTINUE
+        end
+
+        local name, provide, cppfile = compiler_support.get_provided_module(module)
+        cppfile = cppfile or module.cppfile
+
+        local fileconfig = target:fileconfig(cppfile)
+        local bmifile = provide and compiler_support.get_bmi_path(provide.bmi)
+        local build = _should_build(target, cppfile, bmifile, objectfile, module.requires)
+
+        -- add objectfile if module is not from external dep
+        if not (fileconfig and fileconfig.external) then
+            target:add("objectfiles", objectfile)
+        end
+
+        -- needed to detect rebuild of dependencies
+        if provide then
+            compiler_support.memcache():set2(target:name(), name, build)
```

Comment:
we should use a key name prefix, like "should_rebuild_in .. target:name()", and please wrap a function.

```lua
if build then
    _mark_rebuild(name, target)
end
```